new "type system concepts" section (#1743)
python/typing: docs/spec/protocol.rst

cases we can just add a qualifier such as *protocol classes*
when referring to the static type concept.

If a class includes a protocol in its MRO, the class is called an *explicit*

> subclass of the protocol. If a class is a structural consistent subtype of a

Author: Tried some more explicit wording, let me know what you think.

cases we can just add a qualifier such as *protocol classes*
when referring to the static type concept.

If a class includes a protocol in its MRO, the class is called an *explicit*
subclass of the protocol. If a class defines all attributes and methods of a
protocol with types that are assignable to the types of the protocol's
attributes and methods, it is said to implement the protocol and to be
assignable to the protocol. If a class is assignable to a protocol but the
protocol is not included in the MRO, the class is *implicitly* assignable to
the protocol. (Note that one can explicitly subclass a protocol and still not

> implement it if a protocol attribute is set to ``None`` in the subclass, see

Reviewer: ```suggestion
implement it if a protocol attribute is set to ``None`` in the subclass. See
```

Python :py:ref:`data model <specialnames>` for details.)

The attributes (variables and methods) of a protocol that are mandatory for

> another class to be assignable to the protocol are called protocol members.

Reviewer: ```suggestion
another class for it to be assignable to the protocol are called "protocol
members".
```

      def close(self) -> None:
          ...

Now if one defines a class ``Resource`` with a ``close()`` method whose type
signature is assignable to ``SupportsClose.close``, it would implicitly be

> assignable to ``SupportsClose``, since structural assignability is used for

Reviewer: I think this makes it clearer that you're introducing a new term:

```suggestion
assignable to ``SupportsClose``, since "structural assignability" is used for
```

  close_all([f, r])  # OK!
  close_all([1])     # Error: 'int' has no 'close' method

Note that both the user-defined class ``Resource`` and the built-in ``IO`` type

> (the return type of ``open()``) are assignable to ``SupportsClose``, because

Reviewer: Not relevant to this PR's purpose -- but lol, if only `IO` were the [return type of `open()`](https://github.com/python/typeshed/blob/e285e52e1385a3c63a850dd5a2e8b2c744d7293a/stdlib/builtins.pyi#L1527-L1612).

> they provide a ``close()`` method with an assignable type signature.

Reviewer: ```suggestion
each provides a ``close()`` method with an assignable type signature.
```

    represent(nice) # OK
    represent(another) # Also OK

> Note that there is little difference between explicit and implicit consistent

Reviewer 2: Again not sure adding "consistent" here makes sense. Perhaps instead we can talk about "subclasses" of the protocol.

Author: I hadn't meant to leave this open, just got called away suddenly earlier before I'd gotten through all the feedback. I changed this to use similar wording as elsewhere.

  c = One()  # OK
  c = Other()  # Also OK

Assignability relationships with other types
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Protocols cannot be instantiated, so there are no values whose
runtime type is a protocol. For variables and parameters with protocol types,
assignability relationships are subject to the following rules:

* A protocol is never assignable to a concrete type.
* A concrete type ``X`` is assignable to a protocol ``P``
  if and only if ``X`` implements all protocol members of ``P`` with
  assignable types. In other words, assignability with respect to a protocol is

>   always structural.

Reviewer 2: Not necessary with your edit earlier in the file.

  c = One()  # OK
  c = Other()  # Also OK

Assignability relationships with other types
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Protocols cannot be instantiated, so there are no values whose
runtime type is a protocol. For variables and parameters with protocol types,
assignability relationships are subject to the following rules:

> * A protocol is never assignable to a concrete type.

Reviewer 2: Other than `object`?

Author: I think this is a good correction (a protocol type should always be assignable to object).

I'd also like to either define or replace the term "concrete type" in the spec.

But I think both of these can safely be left as a separate follow-up. The "object" change is a clear change to the meaning of the spec that isn't directly tied to this PR. "Concrete type" is more related to this PR in that it's an issue of core terminology, but it's not directly impacted by the terms or concepts introduced in this PR: the term is just as clear/unclear/accurate/inaccurate after this PR as it was before this PR. So we can handle it separately.

Filed https://github.com/python/typing/issues/1781 and https://github.com/python/typing/issues/1780

  if it only contains methods as members (for example ``Sized``,
  ``Iterator``, etc). A protocol that contains at least one non-method member
  (like ``x: int``) is called a data protocol.
* *Unsafe overlap*: A type ``X`` is called unsafely overlapping with a protocol
  ``P``, if ``X`` is not assignable to ``P``, but it is assignable to the type

>   erased version of ``P`` where all members have type ``Any``. In addition, if

Reviewer: I think "type erased" should be hyphenated here
